storage/sql/sql.go

```diff
@@ -59,19 +60,35 @@ var (
 		// NOTE(ericchiang): For some reason using `SET SESSION CHARACTERISTICS AS TRANSACTION` at a
 		// session level didn't work for some edge cases. Might be something worth exploring.
 		executeTx: func(db *sql.DB, fn func(sqlTx *sql.Tx) error) error {
-			tx, err := db.Begin()
-			if err != nil {
-				return err
-			}
-			defer tx.Rollback()
+			ctx, cancel := context.WithCancel(context.TODO())
+			defer cancel()
 
-			if _, err := tx.Exec(`SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;`); err != nil {
-				return err
+			opts := &sql.TxOptions{
+				Isolation: sql.LevelSerializable,
 			}
-			if err := fn(tx); err != nil {
-				return err
+
+			for {
+				tx, err := db.BeginTx(ctx, opts)
+				if err != nil {
+					return err
+				}
+
+				if err := fn(tx); err != nil {
+					return err
```

Comment:
Ack, I think this error also needs to be checked. Our integration tests just failed, so maybe it was luck that it passed for me. I think the `fn` here returns an error because whatever query it's running sees the serialization failure. Unfortunately the `fn` also sometimes wraps the error, making it impossible to check the error code.